ssl/t1_enc.c

```diff
@@ -432,13 +437,36 @@ int tls1_change_cipher_state(SSL *s, int which)
     crypto_info.tls_vmajor = (s->version >> 8) & 0x000000ff;
     crypto_info.tls_vminor = (s->version & 0x000000ff);
 # else
-    /* check that cipher is AES_GCM_128 */
-    if (EVP_CIPHER_nid(c) != NID_aes_128_gcm
-        || EVP_CIPHER_mode(c) != EVP_CIPH_GCM_MODE
-        || EVP_CIPHER_key_length(c) != TLS_CIPHER_AES_GCM_128_KEY_SIZE)
+    /* check that cipher is AES_GCM_128, AES_GCM_256, AES_CCM_128 */
+    switch (EVP_CIPHER_nid(c))
+    {
+#  ifdef OPENSSL_KTLS_AES_GCM_128
+    case NID_aes_128_gcm:
+        if (EVP_CIPHER_mode(c) != EVP_CIPH_GCM_MODE ||
+            EVP_CIPHER_key_length(c) != TLS_CIPHER_AES_GCM_128_KEY_SIZE)
```

Comment:
Do not exactly know about mode checks, but key length checks are made to be sure that we will not overwrite any memory later in memcpy(). This checks were cloned from previous code of kernel TLS support. Additional check of tag len in CCM mode is needed because NID_aes_128_ccm is used for bith AES128-CCM and AES128-CCM8 but latter is not supported in kernel.

Reply:
But both mode and key length are given in the nid! If we get a NID of NID_aes_128_gcm and the EVP_CIPHER_mode() returns anything other than `EVP_CIPH_GCM_MODE` then something has gone fundamentally wrong. Similarly with the key length. This code looks redundant to me. I realise it was there in the original code - but now we're duplicating this same redundant code everywhere, which seems wrong.

Reply:
Tag length checks make sense

Reply:
Redundant checks are removed and the hole check is moved to `ktls_check_supported_cipher` in `ktls.h`